File: src/de/hsh/kevin/logic/myActor/ProjectileManager.java
package de.hsh.kevin.logic.myActor;

import de.hsh.kevin.logic.Config;
import javafx.scene.canvas.Canvas;

import java.util.ArrayList;
import java.util.List;

/**
 * Verwaltet Projektile
 * @author Kevin
 *
 */
public class ProjectileManager {

    private ArrayList<Projectile> projectile;
    private long lastProjectileSpawn;

    /**
     * Erstellt den ProjektilManager ohne Projektile
     */
    public ProjectileManager() {
        projectile = new ArrayList<>();
    }

    /**
     * Liefert alle Projektile
     * @return
     */
    public List<Projectile> getProjectiles() {
        return projectile;
    }

    /**
     * Erstellt ein neues Projektil bei dem Spieler
     * @param playerPos Position des Players wo Projektil erstellt werden soll
     * @param offset zur Position des Players
     */
    public void createProjectile(double[] playerPos, double offset) {

        long curMillis = System.currentTimeMillis();
        if (curMillis >= lastProjectileSpawn + Config.projectileSpawnDelay) {
            Projectile p = ProjectileFactory.getProjectile(playerPos[0] + offset, playerPos[1]);
            projectile.add(p);
            lastProjectileSpawn = curMillis;
        }
    }

    /**
     * Bewegt alle Pakete
     */
    public void move() {
        ArrayList<Projectile> toRemove = new ArrayList<>();

        projectile.stream().parallel().forEach(proj -> {
            if (proj.getPos()[1] < 0) {
                toRemove.add(proj);
            } else {
                proj.move();
            }
        });

        for (int i = 0; i < toRemove.size(); i++) {
            projectile.remove(toRemove.get(i));
        }

        for (int i = 0; i < projectile.size(); i++) {

        }
    }

    /**
     * Zeichnet alle Pakete aufs Canvas
     * @param canvas
     */
    public void draw(Canvas canvas) {
        for (Projectile p : projectile) {
            p.draw(canvas);
        }
    }

    /**
     * Entfernt ein Paket
     * @param p zu entfernendes Paket
     */
    public void remove(Projectile p) {
        projectile.remove(p);
    }

}

File: Games/Kevin/src/de/hsh/kevin/logic/myActor/ProjectileManager.kt
package de.hsh.kevin.logic.myActor

import javafx.scene.canvas.Canvas
import java.util.*

/**
 * Verwaltet Projektile
 * @author Kevin
 */
class ProjectileManager {
    val projectile = ArrayList<Projectile>()
    private var lastProjectileSpawn: Long = 0

    /**
     * Erstellt ein neues Projektil bei dem Spieler
     * @param playerPos Position des Players wo Projektil erstellt werden soll
     * @param offset zur Position des Players
     */
    fun createProjectile(playerPos: DoubleArray, offset: Double) {
        val curMillis = System.currentTimeMillis()
        if (curMillis >= lastProjectileSpawn + 400) {
            projectile.add(Projectile(playerPos[0] + offset, playerPos[1]))
            lastProjectileSpawn = curMillis
        }
    }

    /**
     * Bewegt alle Pakete
     */
    fun move() {
        val toRemove = ArrayList<Projectile>()
        projectile.forEach{ proj: Projectile ->
            if (proj.pos[1] < 0) {
                toRemove.add(proj)
            } else {
                proj.move()
            }
        }
        for (value in toRemove) {
            projectile.remove(value)
        }
    }

    /**
     * Zeichnet alle Pakete aufs Canvas
     * @param canvas
     */
    fun draw(canvas: Canvas) {
        for (p in projectile) {
            p.draw(canvas)
        }
    }

    /**
     * Entfernt ein Paket
     * @param p zu entfernendes Paket
     */
    fun remove(p: Projectile?) {
        projectile.remove(p)
    }
}
